### PR TITLE
added vlogs support

### LINKED ISSUE #1052: Add VLogs to the operator API
# Goals

* Managed vlogs application by operator.
* Simplified deployment and configuration.

# Implementation details

* Use the same logic that we already have for `VMSingle`. 
* Use `Deployment` with restart strategy `Recreate`. 
* move useful configuration flags into `Spec` field. 
* add ingest ports configuration, such as syslog server.

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**
Fixes #1052, #686

### CODE REVIEW

File: config/rbac/kustomization.yaml [hunk 1]
```diff
@@ -13,6 +13,8 @@ resources:
 # default, aiding admins in cluster management. Those roles are
 # not used by the Project itself. You can comment the following lines
 # if you do not want those helpers be installed with your Project.
+- operator_vlogs_editor_role.yaml
```
Comment (f41gh7) [inline config/rbac/kustomization.yaml:16]: looks like it should disabled by default

File: config/rbac/role.yaml [hunk 2]
```diff
@@ -1,187 +1,13 @@
...
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
-  labels:
-    app.kubernetes.io/name: vm-operator
-    app.kubernetes.io/managed-by: kustomize
-  name: operator
+  name: manager-role
 rules:
-- apiGroups:
```
Comment (f41gh7) [inline config/rbac/role.yaml]: Any reason to delete access to other components?

File: internal/controller/operator/factory/vlogs/vlogs.go [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	vmSingleDataDir  = "/victoria-metrics-data"
```
Comment (f41gh7) [inline internal/controller/operator/factory/vlogs/vlogs.go]: WDYT of vlogsDataDir and `/victoria-logs-data`?

File: internal/controller/operator/factory/vlogs/vlogs.go [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+			DNSPolicy:                     r.Spec.DNSPolicy,
+			DNSConfig:                     r.Spec.DNSConfig,
+			TopologySpreadConstraints:     r.Spec.TopologySpreadConstraints,
+			HostAliases:                   r.Spec.HostAliases,
+			TerminationGracePeriodSeconds: r.Spec.TerminationGracePeriodSeconds,
+			ReadinessGates:                r.Spec.ReadinessGates,
+		},
+	}
+
+	return vmSingleSpec, nil
```
Comment (f41gh7) [inline internal/controller/operator/factory/vlogs/vlogs.go]: nit vLogsSpec

File: docs/api.md [hunk 5]
```diff
@@ -2222,6 +2228,83 @@ _Appears in:_
...
+| `imagePullSecrets` | ImagePullSecrets An optional list of references to secrets in the same namespace<br />to use for pulling images from registries<br />see https://kubernetes.io/docs/concepts/containers/images/#referring-to-an-imagepullsecrets-on-a-pod | _[LocalObjectReference](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#localobjectreference-v1-core) array_ | false |
+| `initContainers` | InitContainers allows adding initContainers to the pod definition. Those can be used to e.g.<br />fetch secrets for injection into the vmSingle configuration from external sources. Any<br />errors during the execution of an initContainer will lead to a restart of the Pod. More info: https://kubernetes.io/docs/concepts/workloads/pods/init-containers/<br />Using initContainers for any use case other then secret fetching is entirely outside the scope<br />of what the maintainers will support and by doing so, you accept that this behaviour may break<br />at any time without notice. | _[Container](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#container-v1-core) array_ | false |
```
Comment (f41gh7) [inline docs/api.md]: nit vmSingle-> vlogs

File: docs/api.md [hunk 6]
```diff
@@ -2222,6 +2228,83 @@ _Appears in:_
...
+| `removePvcAfterDelete` | RemovePvcAfterDelete - if true, controller adds ownership to pvc<br />and after VLogs objest deletion - pvc will be garbage collected<br />by controller manager | _boolean_ | false |
+| `replicaCount` | ReplicaCount is the expected size of the VLogs<br />it can be 0 or 1<br />if you need more - use vm cluster | _integer_ | true |
+| `resources` | Resources container resource request and limits, https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/<br />if not defined default resources from operator config will be used | _[ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#resourcerequirements-v1-core)_ | false |
+| `retentionPeriod` | RetentionPeriod for the stored logs<br />Note VictoriaLogs has data/ and indexdb/ folders<br />logs from data/ removed eventually as soon as partition leaves retention period<br />reverse index data at indexdb rotates once at the half of configured [retention period](https://github.com/VictoriaMetrics/VictoriaMetrics/tree/master/docs/Single-server-VictoriaMetrics.md/#retention) | _string_ | true |
```
Comment (f41gh7) [inline docs/api.md]: nit link to the vlogs doc

**Review by f41gh7 — COMMENTED** [review]
Looks good!

- Please add example into the config/examples. It will make it possible to use with operator-hub

- address typos

File: config/rbac/kustomization.yaml (hunk 1 above)
Comment (AndrewChubatiuk) [reply config/rbac/kustomization.yaml:16]: it was removed automatically by kubebuilder, looks like it doesn't like config folder structure

File: config/rbac/role.yaml (hunk 2 above)
Comment (AndrewChubatiuk) [reply config/rbac/role.yaml]: it was removed automatically by kubebuilder, looks like it doesn't like config folder structure

File: internal/controller/operator/factory/vlogs/vlogs.go (hunk 3 above)
Comment (AndrewChubatiuk) [reply internal/controller/operator/factory/vlogs/vlogs.go]: updated

**Review by f41gh7 — APPROVED** [review]
LGTM

**Comment by f41gh7** [comment]
Thanks!

**Pull request merged**